src/persist/src/indexed/runtime.rs

```diff
@@ -0,0 +1,319 @@
+/// Move this work out into whatever async runtime the user likes, via something
+/// like https://docs.rs/rdkafka/0.26.0/rdkafka/util/trait.AsyncRuntime.html
+pub fn start<U: Buffer + Send + 'static, L: Blob + Send + 'static>(
+    indexed: Indexed<U, L>,
+) -> RuntimeClient {
+    let (tx, rx) = mpsc::channel();
+    let handle = thread::spawn(move || {
+        // TODO: Set up the tokio or other async runtime context here.
+        let mut l = RuntimeImpl { indexed, rx };
+        while l.work() {}
+    });
+    let handle = Mutex::new(Some(handle));
+    let core = RuntimeCore { handle, tx };
+    RuntimeClient {
+        core: Arc::new(core),
+    }
+}
+
+/// A receiver for the `Result<T, Error>` response of an asynchronous command.
+pub enum CmdResponse<T> {
+    /// No response is desired.
+    Ignore,
+    /// A single channel send per command call.
+    ///
+    /// It is supported for multiple command results to be sent to the same
+    /// channel by cloning the sender. This even works if they are different
+    /// types of command, but the expected usage is to issue N writes and then
+    /// wait on them all by receiving N times.
+    ///
+    /// NB: The given Sender may be dropped without having received a message if
+    /// the runtime is shut down.
+    Channel(Sender<Result<T, Error>>),
+    /// A callback executed once.
+    ///
+    /// NB: This isn't guaranteed to get called because of a race with runtime
+    /// shutdown.
+    Callback(Box<dyn FnOnce(Result<T, Error>) + Send + 'static>),
+    // TODO: Add a std::future::Future variant to better play with async? I
+    // think the Future impl would be pretty straightforward.
+}
+
+impl<T> From<Sender<Result<T, Error>>> for CmdResponse<T> {
+    fn from(c: Sender<Result<T, Error>>) -> Self {
+        CmdResponse::Channel(c)
+    }
+}
+
+impl<T> From<Box<dyn FnOnce(Result<T, Error>) + Send + 'static>> for CmdResponse<T> {
+    fn from(c: Box<dyn FnOnce(Result<T, Error>) + Send + 'static>) -> Self {
+        CmdResponse::Callback(c)
+    }
+}
+
+impl<T> CmdResponse<T> {
+    fn send(self, t: Result<T, Error>) {
+        match self {
+            CmdResponse::Ignore => {}
+            CmdResponse::Callback(c) => c(t),
+            CmdResponse::Channel(c) => {
+                if let Err(SendError(_)) = c.send(t) {
+                    // The SendError docs indicate that this only happens if the
+                    // receiver was dropped. If the client that sent this cmd
+                    // has hung up (imagine an INSERT spawned a Cmd::Write, but
+                    // the pgwire connection has been severed while it synced to
+                    // disk), then no one is listening and that's okay.
+                }
+                // Defensively drop c, just to make it obvious.
+                drop(c)
+            }
+        }
+    }
+}
+
+struct RuntimeCore {
+    handle: Mutex<Option<JoinHandle<()>>>,
+    tx: mpsc::Sender<Cmd>,
```

Comment:
Hit an issue with this. RuntimeCore is not Sync because of this Sender and as an eventual result, StoragePersister (a thin handle around the runtime, renamed Persister in HEAD but not yet rebased into this branch) is not Sync. I was thinking it would be sufficient for StoragePersister to be Clone+Send, given the requirements of thread::spawn, but it looks like timely::execute also requires the closure to be Sync. Removing the Sync requirement in timely::execute is possible but quite invasive (the Fn would have to be split into a "constructor" Fn that is run in the main thread and returns a "work" Fn that is passed to the spawned thread), so I think it's better just to make this whole "persistence handle" stack Send+Sync.

It seems fine to me for there to be a Mutex lock on register/snapshot/stop but I really don't want one in the write(+seal?) path, but I think that's pretty doable. Rebase on top of main + this change incoming

Reply:
Update: the way I was thinking to do this ended up being workable but gross. @benesch talked me into using crossbeam_channel given that its Senders impl Sync (which makes everything so much less of a hassle) and it beats std::sync::mpsc in basically every way performance-wise. PTAL

https://github.com/crossbeam-rs/crossbeam/tree/311124c8336c80dfdce1b8f4b29f2a27f035725c/crossbeam-channel/benchmarks